feat(scopes): Allow apps to define different API scopes for different…

nickvergessen opened the pull request:
… target clients

## Sibling PRs
- Server PR with `OpenAPI` attribute: https://github.com/nextcloud/server/pull/41175
- Sample PR in notifications: https://github.com/nextcloud/notifications/pull/1707
- Adjust openapi-viewer to show them: https://github.com/nextcloud/ocs_api_viewer/pull/11

## Possible optimisations

- [x] Only put schemas in the openapi-*.jsons when they are used in the scope
- [x] Handle tags 
- [x] Set administration depending on `NoAdminRequired`
- [x] Ensure a method with multiple scopes is added multiple times
- [x] Update docs https://github.com/nextcloud/documentation/pull/11271

Fix #20

nickvergessen commented on generate-spec:
```diff
@@ -412,217 +433,223 @@ foreach ($parsedRoutes as $key => $value) {
 	}
 }
 
-foreach ($routes as $route) {
```
Just had to retab, so better review with https://github.com/nextcloud/openapi-extractor/pull/24/files?w=1

provokateurin commented:
@nickvergessen would it be possible to get this done before the server release so we can update it in the server? This is quite a significant change so it would be great to have it from the start.

provokateurin commented:
The title of the specs should reflect the scope so a short note should be prepended to the current titles.

provokateurin commented on generate-spec:
```diff
@@ -680,18 +714,89 @@ if (count($schemas) == 0 && count($routes) == 0) {
+		}
+
+		// Queue potential sub-refs for exporting as well
+		if (isset($schemas[$schemaName]['allOf'])) {
```
Only checking allOf is not enough. All schemas need to be traversed and checked for a property `$ref` (I think you already realized that with the array items from the last commit)

provokateurin commented on generate-spec:
```diff
@@ -680,18 +714,89 @@ if (count($schemas) == 0 && count($routes) == 0) {
+			foreach ($routeData['responses'] as $statusCode => $responseData) {
+				if (isset($responseData['content']['application/json'])) {
+					$usedSchemas = array_merge($usedSchemas, Helpers::collectUsedRefs($responseData['content']['application/json']['schema']));
+				} elseif (isset($responseData['content']['*/*']['schema']['type'], $responseData['content']['*/*']['schema']['format'])
+					&& $responseData['content']['*/*']['schema']['type'] === 'string'
+					&& $responseData['content']['*/*']['schema']['format'] === 'binary') {
```
This can also happen for other content types, not only `*/*`.

provokateurin commented on src/ControllerMethod.php:
```diff
@@ -84,7 +84,7 @@ static function parse(string $context, array $definitions, ClassMethod $method,
 								$responseDescriptions[$statusCode] = $docNode->value->description;
 							}
-							$responses[] = new ControllerMethodResponse($docNode->value->type, $statusCode, "text/plain", new OpenApiType(type: "string"), null);
+							$responses[] = new ControllerMethodResponse($docNode->value->type, $statusCode, "application/json", new OpenApiType(type: "array", maxLength: 0), null);
```
This is wrong. Throwing OCS*Exception definitely returns text/plain with the error message as a string.

nickvergessen replied:
> This is wrong. Throwing OCS*Exception definitely returns text/plain with the error message as a string.

I added this in a talk controller:
```php
throw new OCSNotFoundException('hi');
```
And the result is:

![grafik](https://github.com/nextcloud/openapi-extractor/assets/213943/c9dcc838-e361-46df-bea1-2769f6167092)

provokateurin replied:
Uhhh, then it's maybe different handling between Controller and OCSController?

nickvergessen replied:
Throwing `OCS*Exception` on NON-ocs controller should be illegal and just ignored and treaten as a normal exception....

I will test it and check.
Any app you have in mind?

nickvergessen replied:
Extracted to https://github.com/nextcloud/openapi-extractor/pull/43

provokateurin commented on generate-spec:
```diff
@@ -680,18 +714,91 @@ if (count($schemas) == 0 && count($routes) == 0) {
+		}
+
+		// Queue potential sub-refs for exporting as well
+		foreach (['allOf', 'oneOf', 'anyOf', 'properties'] as $group) {
```
Still not enough, for example `additionalProperties` can also contain a schema. I would say to be save it is best to really traverse all the schemas otherwise some might be missed.

nickvergessen replied:
looping over all does not work, because `items` e.g. needs to be treated differently then those others. But I checked all from https://github.com/nextcloud/openapi-extractor/blob/a07d21cbc1f731ed4c1359a64e5c79db5203383d/src/OpenApiType.php#L73-L94 now.

provokateurin commented:
Please rebase

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
 	}
+
+	protected static function getScopeNameFromAttributeArgument(Arg $arg, string $routeName): ?string {
+		if ($arg->name->name === 'scope') {
```
The argument could also be unnamed and in that case we'd need to check at which index it is (of the unnamed args).

provokateurin replied:
Additionally I would prefer if the check for the right arg is moved out here and this method only handles the arg where we already know that it is the scope arg.

nickvergessen replied:
> The argument could also be unnamed and in that case we'd need to check at which index it is (of the unnamed args).

I didn't know it's optional, but we could just make it a policy that it's required?

nickvergessen replied:
> Additionally I would prefer if the check for the right arg is moved out here and this method only handles the arg where we already know that it is the scope arg.

Sure can do that, but then we duplicate it or have another method to verify the name, but might make sense with the position thing as well.

provokateurin replied:
You mean by failing if there are unnamed args?

nickvergessen replied:
yeah

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+	protected static function getScopeNameFromAttributeArgument(Arg $arg, string $routeName): ?string {
+		if ($arg->name->name === 'scope') {
+			if ($arg->value instanceof ClassConstFetch) {
+				if ($arg->value->class->getLast() === 'OpenAPI') {
+					return self::getScopeNameFromConst($arg->value);
+				}
```
We limit the types to the consts of the attribute or a string, so we can panic on any other const that is passed

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+			'SCOPE_ADMINISTRATION' => 'administration',
+			'SCOPE_FEDERATION' => 'federation',
+			'SCOPE_IGNORE' => 'ignore',
+			// Fall back for future scopes assuming we follow the pattern (cut of 'SCOPE_' and lower case)
+			default => strtolower(substr($scope->name->name, 6)),
```
I'd rather panic in this case. If there is a new scope then it will also be added to this tool and thus the developers need to update to a newer version.

nickvergessen replied:
Well that is exactly the reason why I would not do it. Maybe updating the tooling is not easily possible or comes with breaking changes in the future but a release needs to be made quickly due to time pressure/major release support/etc.

provokateurin replied:
Alright

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+						continue;
+					}
+
+					$foundsTags = [];
```
```suggestion
					$foundTags = [];
```

nickvergessen replied:
🙈

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+					foreach ($attr->args as $arg) {
+						$foundScopeName = self::getScopeNameFromAttributeArgument($arg, $routeName);
+
+						if ($arg->name->name === 'tags') {
+							if ($arg->value instanceof Array_) {
+								foreach ($arg->value->items as $item) {
+									if ($item instanceof ArrayItem) {
+										if ($item->value instanceof String_) {
+											$foundsTags[] = $item->value->value;
+										}
+									}
+								}
+							}
+						}
```
Please un-nest this and panic in case any of the conditions is false.

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+		return $tags;
+	}
+
+	public static function collectUsedRefs(array $data): array {
```
I fear this could run into and endless loop in case there are cyclic dependencies between schemas (which is not forbidden by json schema). Maybe checking if a ref has already been found and then skipping to check if that ref contains more refs will solve it, although that won't work easily with the recursive approach.

nickvergessen replied:
It can not be infinite, we only operate on the given array and then continue with a smaller subset of the same array.

The point that could have a loop is:
```
while ($usedSchema = array_shift($usedSchemas)) {
```

But this loop is prevented by checking for existance before we add them to the `$usedSchemas`:
https://github.com/nextcloud/openapi-extractor/blob/296b99cc52a4a5bf003cd7bd7e15f196d0b284ee/generate-spec#L820-L822
```
				if (!isset($scopedSchemas[substr($newRef, strlen('#/components/schemas/'))])) {
					$usedSchemas[] = $newRef;
				}
```

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+		};
+	}
+
+	public static function getAttributeScopes(ClassMethod|Class_|Node $node, string $annotation, string $routeName): array {
```
`$annotation` is always `OpenAPI`, it could just be skipped from the parameters (and the method name should reflect that).

nickvergessen replied:
Yeah, just followed the pattern of existing methods, but yeah they are used for different attributes, so we can simplify this

provokateurin commented on generate-spec:
```diff
@@ -302,6 +302,12 @@ foreach ($parsedRoutes as $key => $value) {
 		}
 
+		$controllerScopes = Helpers::getAttributeScopes($controllerClass, 'OpenAPI', $routeName);
+		if (in_array('ignore', $controllerScopes, true)) {
+			Logger::info($routeName, "Controller '" . $controllerName . "' ignored because of OpenAPI attribute");
+			continue;
+		}
```
If there's more than one scope but also the ignore scope, we should probably print an error because it was not intended.

provokateurin commented on generate-spec:
```diff
@@ -355,6 +361,24 @@ foreach ($parsedRoutes as $key => $value) {
 		}
 
+		$scopes = Helpers::getAttributeScopes($classMethod, 'OpenAPI', $routeName);
+		if (in_array('ignore', $scopes, true)) {
+			Logger::info($routeName, "Route ignored because of OpenAPI attribute");
+			continue;
+		}
```
Same as above.

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +162,118 @@ public static function cleanSchemaName(string $name): string {
+		return $scopes;
+	}
+
+	public static function getAttributeTagsByScope(ClassMethod|Class_|Node $node, string $annotation, string $routeName, string $defaultTag, string $defaultScope): array {
```
Here `$annotation` is also always `OpenAPI`

provokateurin commented on generate-spec:
```diff
@@ -712,15 +751,100 @@ if (count($schemas) == 0 && count($routes) == 0) {
+	$openapiScope['paths'] = $paths;
+
+	$usedSchemas = [];
+	foreach ($paths as $url => $urlRoutes) {
+		foreach ($urlRoutes as $httpMethod => $routeData) {
+			foreach ($routeData['responses'] as $statusCode => $responseData) {
+				if (isset($responseData['content']['application/json'])) {
+					if (is_array($responseData['content']['application/json']['schema'])) {
+						$newSchemas = Helpers::collectUsedRefs($responseData['content']['application/json']['schema']);
+						$usedSchemas = array_merge($usedSchemas, $newSchemas);
+					}
+				} elseif (isset($responseData['content']['*/*']['schema']['type'], $responseData['content']['*/*']['schema']['format'])
+					&& $responseData['content']['*/*']['schema']['type'] === 'string'
+					&& $responseData['content']['*/*']['schema']['format'] === 'binary') {
+					Logger::info("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
+				} else {
+					Logger::info("app", "Could not read used schemas for response to '$httpMethod $url' with status code $statusCode");
+				}
+			}
+		}
+	}
```
I feel like this is too specific and can break easily. For example it doesn't account for requestBody (which isn't used yet, but probably will be in the future). I would appreciate a more general approach that just checks everything and not the specific parts of the responses.

nickvergessen replied:
Not sure how to improve this. If something else than JSON/data is returned, there can not be a ref with a schema type in it.

Can you give a sample of what you think a problem is, so we can generate a test case and look at a solution

provokateurin replied:
I would just loop all the content types too and then call collectUsedRefs on the schema. If the type is binary it will find no refs, so there shouldn't be a problem. This way the different content types are handled the same.
There are also more content types than the ones that are used here, in fact any content type can appear since some Responses allow setting it manually.

nickvergessen replied:
I still don't get this. The only place this happens atm is with OCS exceptions, and that is a bug in the tool (handled in #43 ).

I improved the detection and handling in the last commit and now think this is good to go.

provokateurin commented on generate-spec:
```diff
@@ -712,15 +751,100 @@ if (count($schemas) == 0 && count($routes) == 0) {
+				} elseif (isset($responseData['content']['*/*']['schema']['type'], $responseData['content']['*/*']['schema']['format'])
+					&& $responseData['content']['*/*']['schema']['type'] === 'string'
+					&& $responseData['content']['*/*']['schema']['format'] === 'binary') {
+					Logger::info("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
```
```suggestion
					Logger::debug("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
```

provokateurin commented on generate-spec:
```diff
@@ -712,15 +751,100 @@ if (count($schemas) == 0 && count($routes) == 0) {
+					&& $responseData['content']['*/*']['schema']['format'] === 'binary') {
+					Logger::info("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
+				} else {
+					Logger::info("app", "Could not read used schemas for response to '$httpMethod $url' with status code $statusCode");
```
What is the point of this message? Should it tell the developers that something is wrong in their code or should it tell us that there is a programming error in this tool?

nickvergessen replied:
Can be either, none of the known apps trigger this path. But if it's an error in the tool it should not completely break the app and at least generate the remaining bits successfully?

provokateurin replied:
I think it can be completely dropped with my suggestion above.

provokateurin commented on generate-spec:
```diff
@@ -712,15 +751,100 @@ if (count($schemas) == 0 && count($routes) == 0) {
+		}
+
+		// Queue potential sub-refs for exporting as well
+		foreach (['allOf', 'oneOf', 'anyOf', 'properties', 'additionalProperties'] as $group) {
+			if (isset($schemas[$schemaName][$group])) {
+				foreach ($schemas[$schemaName][$group] as $subType) {
+					$newRefs = Helpers::collectUsedRefs($subType);
+					foreach ($newRefs as $newRef) {
+						if (!isset($scopedSchemas[substr($newRef, strlen('#/components/schemas/'))])) {
+							$usedSchemas[] = $newRef;
+						}
+					}
+				}
+			}
+		}
+
+		if (isset($schemas[$schemaName]['items'])) {
+			$newRefs = Helpers::collectUsedRefs($schemas[$schemaName]['items']);
+			foreach ($newRefs as $newRef) {
+				if (!isset($scopedSchemas[substr($newRef, strlen('#/components/schemas/'))])) {
+					$usedSchemas[] = $newRef;
+				}
+			}
+		}
```
Duplicate with the helper method, please refactor

nickvergessen commented:
> IMO the output should be a bit different:
> 
> 1. Always generate a openapi.json file. If there is only one scope used in the app, then everything goes there. If there are multiple scopes in the app, then this spec should contain everything from all scopes at once (so what we have right now).
> 2. Generate openapi-xxx.json files for additional scopes.

I'm not sure I agree with `1.` here. In Talk there will never be a consumer that will implement all the available scopes:
- Recording backend
- Signaling backend
- SIPbridge backend
- Bot
- "Normal user client"

It might be a different story for default + administration, but for the others they just flood the list of tags, schemas and endpoints unnecessarily with information.

> In my mind this makes more sense because we still have the spec that contains everything, but for users who don't need everything they can just us openapi-xxx.json.

The default experience will bring them to the full thing however. Also we would need to introduce yet another scope in talk for "default", or generate additionally openapi-default.json, and I'm not sure that would be self explaining enough for devs to realize they only need to implement that one.

> If you don't agree we should make sure that the merge-specs script can combine multiple of these scoped specs to get the full spec again (actually this should work either way 🤔).

I would favor generating a openapi-full.json or openapi-all.json instead. Be it by merge-specs or by generate-specs

provokateurin commented:
> I would favor generating a openapi-full.json or openapi-all.json instead. Be it by merge-specs or by generate-specs

Sounds good, let's add a openapi-full.json :+1:

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +164,125 @@ public static function cleanSchemaName(string $name): string {
+							continue;
+						}
+
+						if (!$arg->value instanceof Array_) {
+
+							continue;
+						}
```
This should panic instead of ignoring wrong values

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +164,125 @@ public static function cleanSchemaName(string $name): string {
+						}
+
+						foreach ($arg->value->items as $item) {
+							if ($item instanceof ArrayItem && $item->value instanceof String_) {
```
```suggestion
							if ($item?->value instanceof String_) {
```

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +164,125 @@ public static function cleanSchemaName(string $name): string {
+					if (!empty($foundTags)) {
+						$tags[$foundScopeName ?: $defaultScope] = $foundTags;
+					}
+
+
```
Random empty lines

provokateurin commented on src/Helpers.php:
```diff
@@ -156,4 +164,125 @@ public static function cleanSchemaName(string $name): string {
+						}
+
+						if (!$arg->value instanceof Array_) {
+
```
Random empty line

provokateurin commented on generate-spec:
```diff
@@ -712,15 +768,111 @@ if (count($schemas) == 0 && count($routes) == 0) {
+					}
+
+					foreach ($responseData['content'] as $contentType => $contentData) {
+						if ($contentType === 'application/json') {
+							if (isset($contentData['schema']) && is_array($contentData['schema'])) {
+								$newSchemas = Helpers::collectUsedRefs($contentData['schema']);
+								$usedSchemas = array_merge($usedSchemas, $newSchemas);
+							}
+						} elseif (
+							(
+								in_array($contentType, ['*/*', 'text/css', 'application/octet-stream'], true)
+								|| str_starts_with($contentType, 'image/')
+							)
+							&& isset($contentData['schema']['type'], $contentData['schema']['format'])
+							&& $contentData['schema']['type'] === 'string'
+							&& $contentData['schema']['format'] === 'binary') {
+							Logger::debug("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
+						} elseif (isset($contentData['schema'])) {
+							Logger::warning("app", "Could not read used schemas for response to '$httpMethod $url' with status code $statusCode with content type $contentType");
+						}
```
Please remove this logic and just always try collecting refs for the response schema regardless of the content type or schema. We can not be sure that only certain responses contain refs and this code definitely doesn't catch all the possible types because some responses allow setting custom content types for example.

nickvergessen commented:
Rebased again to solve the conflicts from #72

nickvergessen commented on generate-spec:
```diff
@@ -711,17 +767,100 @@ if (count($schemas) == 0 && count($routes) == 0) {
+
+	file_put_contents($scopeOut, json_encode($openapiScope, Helpers::jsonFlags()));
+
+	Logger::info('app', 'Generated scope with' . count($openapiScope['paths']) . ' routes!');
```
ctrl + s fail...

nickvergessen commented:
> Can you update the demo PRs you did so I can review the changes in those and check that everything is good?

* Notifications: was updated already, I rebased it on latest master: https://github.com/nextcloud/notifications/pull/1707
* Spreed: was updated already, I rebased it on latest main (wording changes in -full only): https://github.com/nextcloud/spreed/pull/10856
* Server: at https://github.com/nextcloud/server/pull/42888

provokateurin commented:
Running this in Talk results in
```log
Info: app: Generated scope default with 66 routes!
Info: app: Generated scope bots with 2 routes!
Info: app: Generated scope administration with 9 routes!
Info: app: Generated scope backend-recording with 2 routes!
Info: app: Generated scope backend-sipbridge with 6 routes!
Info: app: Generated scope backend-signaling with 1 routes!
Info: app: Generated scope full with 85 routes!
```
But the sum of all routes is 86 and the full spec has only 85 :thinking:

provokateurin commented:
From the server branch:
https://github.com/nextcloud/server/blob/e1af7061e7c1598e5ccdfcb2fc6144070d0bcef9/apps/settings/openapi.json#L22
Empty schemas, probably missing a stdClass fix (might have been a problem on main already).

nickvergessen commented:
Both fixed and pushed now

provokateurin reviewed (approved):
Noice :tada:
